File: src/sorting/cpp/heap sort/heap_sort.cpp
#include <vector>

template <class T> void heapify(vector<T> &t, int cur, int offset) {

  auto parent = [](int i) { return i >> 1; };

  auto left_child = [](int i) { return i << 1; };

  auto right_child = [](int i) { return (i << 1) + 1; };

  if (cur + offset > t.size())
    return;

  heapify(t, left_child(cur), offset);
  heapify(t, right_child(cur), offset);

  if (cur > 1 && t[cur + offset - 1] < t[parent(cur) + offset - 1])
    swap(t[cur + offset - 1], t[parent(cur) + offset - 1]);
}

template <class T> void heap_sort(vector<T> &t) {
  for (int i = 0; i < t.size(); ++i) {
    heapify<T>(t, 1, i);
  }
}

File: src/sorting/python/heap_sort/src/heap_sort.py
"""Heap sort algorithm implementation."""

from typing import List, TypeVar

T = TypeVar("T")


def heap_sort(array: List[T]) -> List[T]:
    """
    Sort an array using the heap sort algorithm.

    This implementation builds a min-heap and repeatedly extracts the minimum
    element to produce a sorted array in-place.

    Time complexity: O(n log n)
    Space complexity: O(log n) due to recursion stack

    Args:
        array: The list of comparable elements to sort.

    Returns:
        The same list, sorted in ascending order.
    """

    def parent(i: int) -> int:
        return i >> 1

    def left_child(i: int) -> int:
        return i << 1

    def right_child(i: int) -> int:
        return (i << 1) + 1

    def heapify(arr: List[T], curr: int, offset: int) -> None:
        if curr + offset > len(arr):
            return

        heapify(arr, left_child(curr), offset)
        heapify(arr, right_child(curr), offset)

        if curr > 1 and arr[curr + offset - 1] < arr[parent(curr) + offset - 1]:
            arr[curr + offset - 1], arr[parent(curr) + offset - 1] = (
                arr[parent(curr) + offset - 1],
                arr[curr + offset - 1],
            )

    for i in range(len(array)):
        heapify(array, 1, i)

    return array
